lib/libzpool/uzfs_mtree.c

```diff
@@ -302,3 +304,152 @@ uzfs_destroy_txg_diff_tree(void *tree)
 	avl_destroy(temp_tree);
 	umem_free(temp_tree, sizeof (*temp_tree));
 }
+
+void
+uzfs_add_to_incoming_io_tree(zvol_state_t *zv, uint64_t offset, uint64_t len)
+{
+	/*
+	 * Here: Handling of incoming_io_tree creation is for error case only.
+	 *	 It should be handled by replica or caller of uzfs_write_data
+	 */
+	if (!zv->incoming_io_tree)
+		uzfs_create_txg_diff_tree((void **)&zv->incoming_io_tree);
+
+	mutex_enter(&zv->io_tree_mtx);
+	add_to_txg_diff_tree(zv->incoming_io_tree, offset, len);
+	mutex_exit(&zv->io_tree_mtx);
+}
+
+uint32_t
+uzfs_search_incoming_io_tree(zvol_state_t *zv, uint64_t offset, uint64_t len,
+    list_t **list)
+{
+	avl_tree_t *tree = zv->incoming_io_tree;
+	uint32_t count = 0;
+	uzfs_zvol_blk_phy_t *b_entry, *a_entry, *entry;
+	avl_index_t where;
+	uzfs_zvol_blk_phy_t tofind;
+	uint64_t a_end, b_end;
+	list_t *chunk_list;
+	uzfs_io_chunk_list_t  *node;
+
+	if (!tree)
+		return (0);
+
+	chunk_list = umem_alloc(sizeof (*chunk_list), UMEM_NOFAIL);
+	list_create(chunk_list, sizeof (uzfs_io_chunk_list_t),
+	    offsetof(uzfs_io_chunk_list_t, link));
+
+	mutex_enter(&zv->io_tree_mtx);
+
+again:
+	tofind.offset = offset;
+	tofind.len = len;
+
+	// Check for exact match
+	entry = avl_find(tree, &tofind, &where);
+	if (entry) {
+		/*
+		 * Here, added entry length is greater or equals to rebuild
+		 * io len
+		 */
+		if (entry->len >= len)
+			goto done;
+
+		/*
+		 * Here added entry length is smaller than rebuild io len
+		 * so make offset to added offset + length and length to
+		 * len - rebuild io len
+		 */
+		if (entry->len < len) {
+			offset = entry->offset + entry->len;
+			len = len - entry->len;
+			goto again;
+		}
+	}
+
+	/*
+	 * Check for entry whose offset is lesser than to_find.offset
+	 * (or search_entry's offset)
+	 */
+	b_entry = avl_nearest(tree, where, AVL_BEFORE);
+	if (b_entry) {
+		b_end = b_entry->offset + b_entry->len;
+		a_end = offset + len;
+
+		// If b_entry is not overlapping with search_entry
+		if (b_end <= offset)
+			goto after;
+
+		// If b_entry ends after search_entry
+		if (a_end <= b_end) {
+			goto done;
+		}
+
+		/*
+		 * If search_entry ends before b_entry then change
+		 * search_entry's offset to before_entry's end and
+		 * length to length - (overlapping length)
+		 */
+		if (a_end > b_end) {
```

Comment:
(b_entry->offset, offset, b_end, a_end).. actually, 'if' condition is not required.

Reply:
I think it is required, 
if b_entry --- offset:100 and len:100
and search_entry --- offset:130 and len:200
then it will become (100, 130, 200, 330), to remove overlapping of 130 to 200, we need this check, right?
I will also add unit test for all this possibility. since rebuild test is not enough to check all scenario. What you suggest?